xdsl/backend/riscv/register_allocation.py

```diff
@@ -85,38 +111,41 @@ def _free(self, reg: SSAValue) -> None:
         ):
             self.available_registers.push(reg.type)
 
-    def allocate_func(self, func: riscv_func.FuncOp) -> None:
-        for region in func.regions:
-            for block in region.blocks:
-                for op in block.walk_reverse():
-                    # Do not allocate registers on non-RISCV-ops
-                    if not isinstance(op, RISCVOp):
-                        continue
+    def process_riscv_op(self, op: RISCVOp) -> None:
+        for result in op.results:
+            # Allocate registers to result if not already allocated
+            self.allocate(result)
+            # Free the register since the SSA value is created here
+            self._free(result)
```

Comment:
Are you sure this makes sense? Allocating a register and immediately freeing it again? 

I'm not sure I follow what's happening in this function...

Reply:
#1538